electrum/lnpeer.py

```diff
@@ -2750,6 +2750,7 @@ async def wrapped_callback():
                 # return payment_key so this branch will not be executed again
                 return None, payment_key, None
             elif preimage:
+                self.lnworker.maybe_cleanup_mpp(chan.get_scid_or_local_alias(), htlc)
```

Comment:
It is too early to clean-up here. In general we should wait until there has been a revack in both directions.
Specifically, while there is any non-revoked commitment tx (of either party) that contains the htlc, we should not clean-up. This is because if such a commitment tx is broadcast, and the htlc mpp set had resolution ACCEPTED and we have the preimage, then we should use the preimage to claim the htlc onchain.

Reply:
note: as this is about HTLCs we received, it is us who will fulfill or fail them, so maybe the clean-up (which needs to check `chan.hm.is_htlc_irrevocably_removed_yet(ctx_owner=None, ...)` can be done after *we* send revack, as the expected flow is:

```
LOCAL         REMOTE
   ---fulfill -->
   ---  sig  --->
   <---revack ---
   <---  sig  ---
   --- revack--->
*and now we can clean-up the htlc
```